lucene/analysis/icu/src/test/org/apache/lucene/analysis/icu/TestICUTransformCharFilter.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.lucene.analysis.icu;
+
+
+import java.io.IOException;
+import java.io.Reader;
+import java.io.StringReader;
+
+import org.apache.lucene.analysis.Analyzer;
+import org.apache.lucene.analysis.BaseTokenStreamTestCase;
+import org.apache.lucene.analysis.MockTokenizer;
+import org.apache.lucene.analysis.Tokenizer;
+import org.apache.lucene.analysis.core.KeywordTokenizer;
+
+import com.ibm.icu.text.Transliterator;
+import com.ibm.icu.text.UnicodeSet;
+
+
+/**
+ * Test the ICUTransformCharFilter with some basic examples.
+ */
+public class TestICUTransformCharFilter extends BaseTokenStreamTestCase {
+  
+  public void testBasicFunctionality() throws Exception {
+    checkToken(Transliterator.getInstance("Traditional-Simplified"), 
+        "簡化字", "简化字"); 
+    checkToken(Transliterator.getInstance("Katakana-Hiragana"), 
+        "ヒラガナ", "ひらがな");
+    checkToken(Transliterator.getInstance("Fullwidth-Halfwidth"), 
+        "アルアノリウ", "ｱﾙｱﾉﾘｳ");
+    checkToken(Transliterator.getInstance("Any-Latin"), 
+        "Αλφαβητικός Κατάλογος", "Alphabētikós Katálogos");
+    checkToken(Transliterator.getInstance("NFD; [:Nonspacing Mark:] Remove"), 
+        "Alphabētikós Katálogos", "Alphabetikos Katalogos");
+    checkToken(Transliterator.getInstance("Han-Latin"),
+        "中国", "zhōng guó");
+  }
+  
+  public void testRollbackBuffer() throws Exception {
+    checkToken(Transliterator.getInstance("Cyrillic-Latin"),
+        "яяяяя", "âââââ"); // final NFC transform applied
+    checkToken(Transliterator.getInstance("Cyrillic-Latin"), 0,
+        "яяяяя", "a\u0302a\u0302a\u0302a\u0302a\u0302"); // final NFC transform never applied
+    checkToken(Transliterator.getInstance("Cyrillic-Latin"), 2,
```

Comment:
Can you explain why the rollback buffer needs to be so large to correctly handle this transliteration? Isn't it just a chartacter-by-character process?

Reply:
This happens because "Cyrillic-Latin" is a compound transliterator, composed of child transliterators "NFD, [main Cyrillic-Latin tranliteration], NFC". After the inner child transliterator changes a single "я" into "â" (decomposed), the final NFC transformation sees the two-character decomposition of "â", but doesn't do anything "yet" because it's not sure whether it will see more combining diacritics. But NFC never sees those input characters again, because on subsequent invocation, they are excluded by the toplevel filters on the main "Cyrillic-Latin" transliterator (which only pays attention to Cyrillic input characters). So the partially-transliterated, decomposed characters are passed through to the output. There's a solid, more general explanation in the code for ICU [Transliterator](https://github.com/unicode-org/icu/blob/a075ac9c/icu4j/main/classes/translit/src/com/ibm/icu/text/Transliterator.java#L1137-L1163).